### PR TITLE
SupplyChain object support in Prebid 

### PR DESCRIPTION
**Pull request by pm-harshad-mane**
## Type of change
- [X] Feature
- [X] Code style update (formatting, local variables)

## Description of change
- Supply chain object support in Prebid #3870 adapterManager will validate the schain object and will pass it to bidders/adapters only if it is valid, thus bidders/adapters will not require to validate schain object.
- schain param support in PubMatic adapter
- other minor changes in PubMatic adapter for better readability

```
             pbjs.setConfig({
                    "schain": {
                        "ver":"1.0",
                        "complete": 1,
                        "nodes": [
                            {
                                "asi":"indirectseller.com",
                                "sid":"00001",
                                "hp":1
                            }
                        ]     
                    }
                });
```

### CODE REVIEW

**Comment by pm-harshad-mane** [comment]
Hello @jsnellbaker , @bretg 
Could you please review the code?

**Comment by pm-harshad-mane** [comment]
Please do not merge these changes as QA is pending.

File: src/schain.js [hunk 1]
```diff
@@ -0,0 +1,112 @@
+import { isNumber, isStr, isArray, isPlainObject, hasOwn, logError, isInteger } from './utils';
```
Comment (mkendall07) [inline src/schain.js]: should this be part of core or a separate module?

Comment (pm-harshad-mane) [reply src/schain.js]: Agree @mkendall07 , i will move schain as a separate module as all pubs may not need it.

**Comment by pm-harshad-mane** [comment]
Hello @bretg , @jaiminpanchal27 
Can you please help me find why CI job is failing, I checked the logs but I am clueless to find a root-cause.

File: src/schain.js (hunk 1 above)
Comment (pm-harshad-mane) [reply src/schain.js]: Moved the code as a module. Resolving the conversation

**Comment by pm-harshad-mane** [comment]
Thanks @jaiminpanchal27 , CI worked 👍

**Comment by pm-harshad-mane** [comment]
Hello @bretg and @jaiminpanchal27 ,

Could you please review this PR?

File: modules/pubmaticBidAdapter.js [hunk 2]
```diff
@@ -738,6 +738,38 @@ function _blockedIabCategoriesValidation(payload, blockedIabCategories) {
...
+      }
+      if (dctrArr.length > 1) {
+        utils.logWarn(LOG_WARN_PREFIX + 'dctr value found in more than 1 adunits. Value from 1st adunit will be picked. Ignoring values from subsequent adunits');
+      }
+    } else {
+      utils.logWarn(LOG_WARN_PREFIX + 'dctr value not found in 1st adunit, ignoring values from subsequent adunits');
+    }
+  }
+}
+
```
Comment (briandamaged) [inline modules/pubmaticBidAdapter.js]: I'm not quite clear on how this pertains to Supply Chain.  From what I can tell, this is Deal Targeting, which seems like it should probably be moved into a separate PR.

File: modules/schain.js [hunk 3]
```diff
@@ -0,0 +1,127 @@
+import {config} from '../src/config';
+import {getGlobal} from '../src/prebidGlobal';
+import { isNumber, isStr, isArray, isPlainObject, hasOwn, logError, isInteger } from '../src/utils';
+
+// validate the supply chanin object
```
Comment (briandamaged) [inline modules/schain.js]: Typo: "supply chain"

File: modules/pubmaticBidAdapter.js [hunk 4]
```diff
@@ -818,6 +849,7 @@ export const spec = {
       if (impObj) {
         payload.imp.push(impObj);
       }
+      schainConfig = schainConfig || bid.schain;
```
Comment (briandamaged) [inline modules/pubmaticBidAdapter.js]: This line seems a bit strange to me.

According to the implementation of the `copySchainObjectInAdunits(..)` function, it seems like you would always have one of the following two cases:

- The Supply Chain object has been injected into _all_ bids
- The Supply Chain object has been injected into _none of_ the bids

However, this specific line of code suggests that there could be scenarios where some bids of a Supply Chain, and some do not.  Furthermore, if a Bid is missing a Supply Chain, then it's somehow acceptable to copy the Supply Chain from the previous Bid...?

This is also making me question whether or not the Supply Chain should actually be injected into individual bids.  It seems like "Supply Chain" is a property of the request as a whole.

Comment (briandamaged) [reply modules/pubmaticBidAdapter.js]: Coming back to this: it seems like the Supply Chain should be injected into the `bidderRequest` rather than the individual `validBidRequests`.  This would eliminate the need for this strange line of code.

Comment (briandamaged) [reply modules/pubmaticBidAdapter.js]: (To be fair: I'm not actually super familiar w/ the Prebid.js API since I haven't worked on this codebase before.  So, `bidderRequest` might not be serving the purpose that I think it serves.  However, my main point is that it seems like the Supply Chain should be a property of the request/transaction as a whole, rather than a property of the individual Bids)

File: modules/pubmaticBidAdapter.js (hunk 2 above)
Comment (pm-harshad-mane) [reply modules/pubmaticBidAdapter.js]: Yes, this code is not related to schain. 
This is an existing code which is moved into a separate function for the sake of readability

File: modules/schain.md [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+});
+```
+
+## Workflow
+The schain module is not enabled by default as it may not be neccessary for all publishers.
+If required, schain module can be included as following
+```
+    $ gulp build --modules=schain,pubmaticBidAdapter,openxBidAdapter,rubiconBidAdapter,sovrnBidAdapter
+```
+The schian module will validate the schain object passed using pbjs.setConfig API.
```
[the remark this answers is not in the document]
Comment (pm-harshad-mane) [reply modules/schain.md]: fixed :)

File: modules/schain.js (hunk 3 above)
Comment (pm-harshad-mane) [reply modules/schain.js]: fixed :)

File: modules/pubmaticBidAdapter.js (hunk 4 above)
Comment (pm-harshad-mane) [reply modules/pubmaticBidAdapter.js]: I also agree that we should not duplicate the schain data in bids.
I am not sure how we can pass schain data in bidderRequest object through an optional module, this can be done easily from adapterManager.

**Comment by pm-harshad-mane** [comment]
Hello @idettman ,

Could you please review the changes?

File: modules/schain.js [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+  if (!isNumber(schainObject.complete) || !isInteger(schainObject.complete)) {
+    logError(schainErrorPrefix + `schain.complete` + shouldBeAnInteger);
+    return false;
+  }
+
+  // ver: String
+  if (!isStr(schainObject.ver)) {
+    logError(schainErrorPrefix + `schain.ver` + shouldBeAString);
+    return false;
+  }
```
[the remark this answers is not in the document]
Comment (mike-chowla) [reply modules/schain.js]: I had asked @pm-harshad-mane to put the validation logic in.  Considering we're facing a deadline from some DSPs to complete the schain work, I'm wondering if the better option is just to remove the validation logic and let bidders validate the schain objects themselves if they wish to.

**Comment by pm-harshad-mane** [comment]
Hello @idettman and @jaiminpanchal27 ,

Could you please review the PR?

File: modules/schain.js [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    adUnit.bids.forEach(bid => {
+      bid.schain = schainObject;
+    });
+  });
+}
+
+export function init(config) {
+  getGlobal().requestBids.before(function(fn, reqBidsConfigObj) {
+    let schainObject = config.getConfig('schain');
+    if (isSchainObjectValid(schainObject)) {
```
[the remark this answers is not in the document]
Comment (pm-harshad-mane) [reply modules/schain.js]: As far as I know, the value will be dynamic.

File: test/spec/modules/schain_spec.js [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+    delete schainConfig.nodes[0].ext; // undefined // param is optional thus this will result true
+    expect(isSchainObjectValid(schainConfig)).to.true;
+    schainConfig.nodes[0].ext = true; // boolean
+    expect(isSchainObjectValid(schainConfig)).to.false;
+    schainConfig.nodes[0].ext = []; // array
+    expect(isSchainObjectValid(schainConfig)).to.false;
+  });
+});
+
+describe('', function() {
```
Comment (idettman) [inline test/spec/modules/schain_spec.js]: Try to use something descriptive rather than empty quotes

Comment (pm-harshad-mane) [reply test/spec/modules/schain_spec.js]: ohh my bad, i will fix it.

File: modules/widespaceBidAdapter.js [hunk 9]
```diff
@@ -177,7 +177,10 @@ export const spec = {
 function storeData(data, name, stringify = true) {
   const value = stringify ? JSON.stringify(data) : data;
   if (LOCAL_STORAGE_AVAILABLE) {
-    localStorage.setItem(name, value);
+    try {
```
Comment (jaiminpanchal27) [inline modules/widespaceBidAdapter.js]: Lets do this in separate PR and tag widespace contributors so they are aware of this change

Comment (pm-harshad-mane) [reply modules/widespaceBidAdapter.js]: ok, I will revert this change

Comment (pm-harshad-mane) [reply modules/widespaceBidAdapter.js]: Reverted the commit.

**Comment by pm-harshad-mane** [comment]
Hello @mkendall07 ,

Can you please review the new changes?

**Comment by pm-harshad-mane** [comment]
Thank you @jaiminpanchal27 , @idettman , @briandamaged and @mkendall07 👍

**Comment by bretg** [comment]
Docs PR https://github.com/prebid/prebid.github.io/pull/1463
